apps/prairielearn/src/ee/lib/validateHTML.ts

```diff
@@ -0,0 +1,602 @@
+        case 'remove-spaces':
+        case 'allow-blank':
+        case 'ignore-case':
+        case 'normalize-to-ascii':
+        case 'show-help-text':
+          assertBool('pl-string-input', key, val, errors);
+          break;
+        default:
+          if (!htmlGlobalAttributes.includes(attr.name)) {
+            errors.push(`pl-string-input: ${attr.name} is not a valid property.`);
+          }
+      }
+    }
+  }
+  if (usedAnswersName === usedCorrectAnswer) {
+    errors.push('pl-string-input: exactly one of answers-name and correct-answer should be set.');
+  }
+  return errors;
+}
+
+/**
+ * Checks that a `pl-checkbox` element has valid properties.
+ * @param ast The tree to consider, rooted at the tag to consider.
+ * @returns The list of errors for the tag, if any.
+ */
+function checkCheckbox(ast: DocumentFragment | ChildNode): string[] {
+  const errors: string[] = [];
+  let usedAnswersName = false;
+  let usedPartialCredit = true;
+  let usedPartialCreditMethod = false;
+  if ('attrs' in ast) {
+    for (const attr of ast.attrs) {
+      const key = attr.name;
+      const val = attr.value;
+      switch (key) {
+        case 'answers-name':
+          usedAnswersName = true;
+          break;
+        case 'weight':
+        case 'number-answers':
+        case 'min-correct':
+        case 'max-correct':
+        case 'min-select':
+        case 'max-select':
+          assertInt('pl-checkbox', key, val, errors);
+          break;
+        case 'inline':
+        case 'fixed-order':
+        case 'hide-help-text':
+        case 'detailed-help-text':
+        case 'hide-answer-panel':
+        case 'hide-score-badge':
+          assertBool('pl-checkbox', key, val, errors);
+          break;
+
+        case 'partial-credit':
+          assertBool('pl-checkbox', key, val, errors);
+          if (
+            ['false', 'f', '0', 'False', 'F', 'FALSE', 'no', 'n', 'No', 'N', 'NO'].includes(val)
+          ) {
+            usedPartialCredit = false;
+          }
+          break;
+        case 'partial-credit-method':
+          assertInChoices('pl-checkbox', key, val, ['COV', 'EDC', 'PC'], errors);
+          usedPartialCreditMethod = true;
+          break;
+        default:
+          if (!htmlGlobalAttributes.includes(key)) {
+            errors.push(`pl-checkbox: ${key} is not a valid property.`);
+          }
+      }
+    }
+  }
+  if (!usedAnswersName) {
+    errors.push('pl-checkbox: answers-name is a required property.');
+  }
+  if (usedPartialCreditMethod && !usedPartialCredit) {
+    errors.push(
+      'pl-checkbox: if partial-credit-method is set, then partial-credit must be set to true.',
+    );
+  }
+
+  let errorsChildren: string[] = [];
+  if ('childNodes' in ast) {
+    for (const child of ast.childNodes) {
+      if ('tagName' in child && child.tagName) {
+        if (child.tagName === 'pl-answer') {
+          errorsChildren = errorsChildren.concat(checkAnswerMultipleChoice(child));
```

Comment:
Using `checkAnswerMultipleChoice` to validate `<pl-answer>` children of `<pl-checkbox>` will allow a `score` attribute that will be rejected at runtime.

Reply:
Changed in 8f5c2be88d306dec842f62a968bf5c24938a0cec.